connect/runtime/src/main/java/org/apache/kafka/connect/storage/OffsetStorageWriter.java

```diff
@@ -98,6 +104,24 @@ private boolean flushing() {
         return toFlush != null;
     }
 
+    public boolean waitForBeginFlush(Supplier<Long> timeout, TimeUnit timeUnit) throws InterruptedException, TimeoutException {
```

Comment:
It seems like we're bending over backwards here to accommodate an assumption made in `beginFlush` that we'll never try to trigger two offset flushes at once, which is clearly false given the conditions that necessitate this fix (i.e., a task's end-of-life offset flush is triggered at the same time as its periodic offset flush).

Given that, do we really need a separate method here, or can we relax the constraints in `beginFlush` to wait for in-progress flushes to conclude instead of throwing an exception if there are any?

Additionally, it seems like the use of a `CompleteableFuture` here is a bit strange. Would a `Semaphore` or `CountDownLatch` be more suited?

Finally--since this change may lead to us performing double offset commits when a task is being shut down, do you think it might also make sense to add a `close` method to the offset writer that throws an exception for any further attempts to flush, and possibly forcibly terminates any in-progress flushes? We can invoke that in `AbstractWorkerTask::cancel` (or possibly `WorkerSourceTask::cancel` if a different approach is necessary to preserve exactly-once semantics) to help tasks complete shutdown within the timeout allotted to them.

Reply:
> Given that, do we really need a separate method here, or can we relax the constraints in beginFlush to wait for in-progress flushes to conclude instead of throwing an exception if there are any?

The reason I implemented it as two separate methods was to minimize the disturbance to other call-sites, specifically the ExactlyOnceSourceTask. Because this introduces a new way to fail due to a timeout, and the ExactlyOnceSourceTask doesn't respect the flush timeout, I thought that it might not be a desirable change where the assertion does no harm.

Upon reflection, I think diverging the usage patterns of the OffsetStorageWriter in these two different contexts is more error-prone than migrating both of them to the new semantics, so I'll update the PR to incorporate your feedback.

> Would a Semaphore or CountDownLatch be more suited?

Thanks, that makes a lot more sense. I was originally trying to re-use the doFlush future, but since `flushing()` starts in `beginFlush`, i needed a synchronizer that i created in beginFlush.

> Finally--since this change may lead to us performing double offset commits when a task is being shut down, do you think it might also make sense to add a close method to the offset writer that throws an exception for any further attempts to flush, and possibly forcibly terminates any in-progress flushes? We can invoke that in AbstractWorkerTask::cancel (or possibly WorkerSourceTask::cancel if a different approach is necessary to preserve exactly-once semantics) to help tasks complete shutdown within the timeout allotted to them.

In EOS mode, if the task is cancelled before it comes time to perform the final offset commit, the final offset commit is skipped. This does not change in this PR, because the ExactlyOnceSourceTask should never leave a flush open and have to wait for a previous flush to finish.

In non-EOS mode, the check of the cancelled flag isn't present. It appears that there are already some wait conditions (in WorkerSourceTask::finalOffsetCommit) to maximize the number of records included in the flush, and that wait condition may cause the thread to commit offsets after cancellation. I don't think that this PR makes double commits possible where they weren't before.

WDYT about adding the EOS-style cancellation semantics to the final commit, or closing the OffsetBackingStore in cancel() to address these cases? Do you think that we can explore those changes in a follow-up PR?

Reply:
> I don't think that this PR makes double commits possible where they weren't before.

So the issue with double commits in non-EOS mode is that, right now, we may throw an exception because of the bug that we're addressing here. But if we fix that exception, then double commits become possible. And if the first commit takes a while, then we might end up lagging too much and performing our second commit after a new instance of the same source task has been brought up.

> WDYT about adding the EOS-style cancellation semantics to the final commit, or closing the OffsetBackingStore in cancel() to address these cases? Do you think that we can explore those changes in a follow-up PR?

I think adding the EOS-style cancellation semantics would be okay for now, though they aren't as effective for this kind of task since we don't have a way of fencing out producers. We can do that part in this PR, and then file a Jira ticket to improve cancellation logic for source task offset commit, which we can explore at a later point.